service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CopyBigQueryDatasetDefinitionStep.java

```diff
@@ -0,0 +1,146 @@
+    if (CloningInstructions.COPY_NOTHING.equals(effectiveCloningInstructions)
+        || CloningInstructions.COPY_REFERENCE.equals(effectiveCloningInstructions)) {
+      // nothing further to do here or on following steps
+      // Build an empty response object
+      final ApiClonedControlledGcpBigQueryDataset result =
+          new ApiClonedControlledGcpBigQueryDataset()
+              .dataset(null)
+              .sourceWorkspaceId(sourceDataset.getWorkspaceId())
+              .sourceResourceId(sourceDataset.getResourceId())
+              .effectiveCloningInstructions(effectiveCloningInstructions.toApiModel());
+      FlightUtils.setResponse(flightContext, result, HttpStatus.OK);
+      return StepResult.getStepResultSuccess();
+    }
+    final String resourceName =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            ControlledResourceKeys.RESOURCE_NAME,
+            ControlledResourceKeys.PREVIOUS_RESOURCE_NAME,
+            String.class);
+    final String description =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            ControlledResourceKeys.RESOURCE_DESCRIPTION,
+            ControlledResourceKeys.PREVIOUS_RESOURCE_DESCRIPTION,
+            String.class);
+    final String datasetName =
+        Optional.ofNullable(
+                inputParameters.get(ControlledResourceKeys.DESTINATION_DATASET_NAME, String.class))
+            .orElseGet(this::randomDatasetName);
+    workingMap.put(ControlledResourceKeys.DESTINATION_DATASET_NAME, datasetName);
+    final UUID destinationWorkspaceId =
+        inputParameters.get(ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    final String location =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            ControlledResourceKeys.LOCATION,
+            ControlledResourceKeys.LOCATION,
+            String.class);
+    final ControlledBigQueryDatasetResource destinationResource =
+        ControlledBigQueryDatasetResource.builder()
+            .accessScope(sourceDataset.getAccessScope())
+            .assignedUser(sourceDataset.getAssignedUser().orElse(null))
+            .cloningInstructions(sourceDataset.getCloningInstructions())
+            .datasetName(datasetName)
+            .description(description)
+            .managedBy(sourceDataset.getManagedBy())
+            .name(resourceName)
+            .resourceId(UUID.randomUUID())
+            .workspaceId(destinationWorkspaceId)
+            .build();
+
+    final ApiGcpBigQueryDatasetCreationParameters creationParameters =
+        new ApiGcpBigQueryDatasetCreationParameters().datasetId(datasetName).location(location);
+    final List<ControlledResourceIamRole> iamRoles =
+        IamRoleUtils.getIamRolesForAccessScope(destinationResource.getAccessScope());
+
+    final ControlledBigQueryDatasetResource clonedResource =
+        controlledResourceService.createBigQueryDataset(
```

Comment:
What happens if the pod this is running on fails after we launch this nested flight? The nested flight might be running on some other pod, so it might continue. This flight will be restarted, this step will be restarted, and we will rerun the nested flight. And the second flight will probably fail, because the first flight finished.

I think to run nested flights, we cannot use the sync service endpoint. We need to be able to do something like:
- in some previous step (or as an input param), make the flightId for the nested flight.
- in this step, test to see if that flight exists. If not, launch it. Either way, wait for completion.

This may be a problem in the other nested flight cases. They are hard to catch in review, because the service method hides the flight inside.

I was discussing this with the Jades the other day in the #stairway channel. Instead of using a nested flight, another approach would be to make a flight step builder method that populates the shared flight steps. So you don't have nesting of the actual flight. You just modularize generating the steps in the one flight. It would get called from both Flight constructors to fill in the "create the resource" steps.

Reply:
We can work this through in the context of PF-918.